### PR TITLE
NEW: Add MapDe architecture.

### PR DESCRIPTION
**Pull request by shaneahmed**
- Add MapDe architecture.

### CODE REVIEW

File: tiatoolbox/models/architecture/micronet.py [hunk 1]
```diff
@@ -276,7 +276,9 @@ def group3_arch_branch(in_ch, skip, out_ch):
     return nn.ModuleDict(module_dict)
 
 
-def group4_arch_branch(in_ch, out_ch, up_kernel=(2, 2), up_strides=(2, 2)):
+def group4_arch_branch(
+    in_ch, out_ch, up_kernel=(2, 2), up_strides=(2, 2), activation="tanh"
+):
     """Group4 branch for MicroNet.
```
Comment (John-P) [inline tiatoolbox/models/architecture/micronet.py:282]: Is it possible to explain a bit more about this? E.g. some more detailed description of what this branch is or where to go to find how it fits into the architecture, e.g. a paper or diagram etc.

File: tiatoolbox/models/architecture/micronet.py [hunk 2]
```diff
@@ -276,7 +276,9 @@ def group3_arch_branch(in_ch, skip, out_ch):
     return nn.ModuleDict(module_dict)
 
 
-def group4_arch_branch(in_ch, out_ch, up_kernel=(2, 2), up_strides=(2, 2)):
+def group4_arch_branch(
+    in_ch, out_ch, up_kernel=(2, 2), up_strides=(2, 2), activation="tanh"
```
Comment (John-P) [inline tiatoolbox/models/architecture/micronet.py:280]:
```suggestion
    in_ch: int, out_ch: int, up_kernel:Tuple[int, int]=(2, 2), up_strides:Tuple[int, int]=(2, 2), activation="tanh"
```

File: tiatoolbox/models/architecture/micronet.py [hunk 3]
```diff
@@ -276,7 +276,9 @@ def group3_arch_branch(in_ch, skip, out_ch):
...
+):
```
Comment (John-P) [inline tiatoolbox/models/architecture/micronet.py:281]:
```suggestion
) -> nn.ModuleDict:
```

File: tiatoolbox/models/architecture/mapde.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+         - Precision
+         - Recall
+         - F1Score
+       * - mapde-crchisto
+         - CRCHisto
+         - 0.81
+         - 0.82
+         - 0.81
+       * - mapde-conic
+         - CoNiC
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/mapde.py:51]:
```suggestion
         - CoNIC
```

File: tiatoolbox/models/architecture/micronet.py (hunk 1 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/micronet.py:291]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/443/commits/4dcf9d765d82e7918ff026fd60b48af65cbb89da

File: tiatoolbox/models/architecture/micronet.py (hunk 2 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/micronet.py]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/443/commits/4dcf9d765d82e7918ff026fd60b48af65cbb89da

File: tiatoolbox/models/architecture/micronet.py (hunk 3 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/micronet.py]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/443/commits/4dcf9d765d82e7918ff026fd60b48af65cbb89da

File: tiatoolbox/models/architecture/mapde.py (hunk 4 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/mapde.py]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/443/commits/4dcf9d765d82e7918ff026fd60b48af65cbb89da

File: tiatoolbox/models/architecture/mapde.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+        coordinates = peak_local_max(
+            np.squeeze(prediction_map[0], axis=2),
+            min_distance=self.min_distance,
+            threshold_abs=self.threshold_abs,
+            exclude_border=False,
+        )
+        return np.fliplr(coordinates)
+
+    @staticmethod
+    def infer_batch(model, batch_data, on_gpu):
```
Comment (John-P) [inline tiatoolbox/models/architecture/mapde.py:255]: Do we need type hints here?

File: tiatoolbox/models/architecture/mapde.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+        """
+        patch_imgs = batch_data
+
+        device = select_device(on_gpu)
+        patch_imgs_gpu = patch_imgs.to(device).type(torch.float32)  # to NCHW
+        patch_imgs_gpu = patch_imgs_gpu.permute(0, 3, 1, 2).contiguous()
+
+        model.eval()  # infer mode
+
+        # --------------------------------------------------------------
```
Comment (John-P) [inline tiatoolbox/models/architecture/mapde.py:285]: What does this line mean? Is it to mark the start of the inference?

**Review by John-P — APPROVED** [review]

File: tiatoolbox/models/architecture/mapde.py (hunk 5 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/mapde.py]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/443/commits/b25b74c348f761586e459e00f51a2169b66edb8f

File: tiatoolbox/models/architecture/mapde.py (hunk 6 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/mapde.py]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/443/commits/b25b74c348f761586e459e00f51a2169b66edb8f

File: tiatoolbox/models/architecture/mapde.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
+"""Defines MapDe architecture.
+
+Raza, Shan E. Ahmed, et al. "Deconvolving convolutional neural network
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/mapde.py:3]:
```suggestion
Raza, Shan E Ahmed, et al. "Deconvolving convolutional neural network
```

File: tiatoolbox/data/pretrained_model.yaml [hunk 8]
```diff
@@ -771,3 +771,23 @@ micronet-consep:
             patch_output_shape: [252, 252]
             stride_shape: [150, 150]
             save_resolution: {'units': 'mpp', 'resolution': 0.25}
+
+mapde-crchisto:
+    url: https://tiatoolbox.dcs.warwick.ac.uk/models/detection/mapde-crchisto.pth
+    architecture:
+        class: mapde.MapDe
```
Comment (mostafajahanifar) [inline tiatoolbox/data/pretrained_model.yaml:778]: You're most probably thinking of adding the `input_resolutions` information here after your have designed the engine. If not, consider adding them based on your training here.

File: tiatoolbox/models/architecture/mapde.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+            input_tensor (torch.Tensor):
+                Input images, the tensor is in the shape of NCHW.
+
+        Returns:
+            torch.Tensor:
+                Output map for cell detection. Peak detection should be applied
+                to this output for cell detection.
+
+        """
+        logits, _, _, _ = super().forward(input_tensor)
```
Comment (mostafajahanifar) [inline tiatoolbox/models/architecture/mapde.py:226]: I see no `preproc` function used here nor in `micronet` implementation. Have you checked the sanity of the outputs that this architecture generates? Because, as far as I remember, you used image normalization in the input of the model also see you have `preproc` method defined in the `micronet` class but never have used these. 

I would suggest you move the required preprocessing of your model into the `forward` method and name the preprocessing method after unet convention to `_transform` [(see here)](https://github.com/TissueImageAnalytics/tiatoolbox/blob/fd957affe0a6cb08fce3eda1431e5812dacb065a/tiatoolbox/models/architecture/unet.py#L324)

File: tiatoolbox/models/architecture/micronet.py [hunk 10]
```diff
@@ -309,25 +330,33 @@ def group4_arch_branch(in_ch, out_ch, up_kernel=(2, 2), up_strides=(2, 2)):
...
             Number of output channels. default=2.
+        activation (str):
+            Activation function, default="softmax".
 
     Returns:
         torch.nn.Sequential:
             An output of type :class:`torch.nn.Sequential`
 
     """
+    if activation == "relu":
```
Comment (mostafajahanifar) [inline tiatoolbox/models/architecture/micronet.py:356]: it seems that you only accept "relu" and "softmax" activations, so I would suggest you check the input for these two types instead of `if...else...` statement.

File: tiatoolbox/models/architecture/mapde.py (hunk 9 above)
Comment (mostafajahanifar) [reply tiatoolbox/models/architecture/mapde.py:226]: Now I can see from the tests that you call preprocessing separately before calling model forward, however, the convention with pytorch model is to include the preprocessing in the forward method.

**Review by mostafajahanifar — CHANGES_REQUESTED** [review]
Thanks Shan for doing this. All looks great, well done.

I have asked for some small changes to make to code more inline with our models' conventions. However, please ignore them if you find the impossible to accommodate.

File: tiatoolbox/data/pretrained_model.yaml (hunk 8 above)
Comment (shaneahmed) [reply tiatoolbox/data/pretrained_model.yaml:778]: I will add them while adding the engine.

File: tiatoolbox/models/architecture/mapde.py (hunk 9 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/mapde.py:226]: Actually there should be pre-proc and that includes subtracting mean and std.
